edge-modules/MetricsCollector/Program.cs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace MetricsCollector
+{
+    using System;
+    using System.Linq;
+    using System.Runtime.Loader;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Client;
+    using Microsoft.Azure.Devices.Client.Transport.Mqtt;
+    using Microsoft.Azure.Devices.Edge.ModuleUtil;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Shared;
+    using Microsoft.Extensions.Logging;
+    using Newtonsoft.Json;
+
+    internal class Program
+    {
+        static readonly Version ExpectedSchemaVersion = new Version("1.0");
+        static readonly ILogger Logger = ModuleUtil.CreateLogger("MetricsCollector");
+        static Timer scrapingTimer;
+
+        public static int Main() => MainAsync().Result;
+
+        private static async Task<int> MainAsync()
+        {
+            Logger.LogInformation($"Starting metrics collector with the following settings:\r\n{Settings.Current}");
+
+            await Init();
+
+            (CancellationTokenSource cts, ManualResetEventSlim completed, Option<object> handler) = ShutdownHandler.Init(TimeSpan.FromSeconds(5), Logger);
+
+            // Wait until the app unloads or is cancelled
+            AssemblyLoadContext.Default.Unloading += ctx => cts.Cancel();
+            Console.CancelKeyPress += (sender, cpe) => cts.Cancel();
+            await WhenCancelled(cts.Token, completed, handler);
+            Logger.LogInformation("MetricsCollector Main() finished.");
+            return 0;
+        }
+
+        /// <summary>
+        ///     Handles cleanup operations when app is cancelled or unloads
+        /// </summary>
+        public static Task WhenCancelled(CancellationToken cancellationToken, ManualResetEventSlim completed, Option<object> handler)

```

Comment:
```suggestion
        public static Task WhenCancelledAsync(CancellationToken cancellationToken, ManualResetEventSlim completed, Option<object> handler)
```

Reply:
This method runs synchronously though, right?